File: openmp/matmul.c
// Matrix multiplication, parallelized using OpenMP.

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

typedef struct
{
    int dimension;
    double *data;
} SquareMatrix;

void matrix_initialize(SquareMatrix *matrix, int dimension)
{
    assert(matrix && "Must be given a pointer to a valid square matrix structure");
    assert(dimension > 0 && "Dimension must be strictly positive integer");

    matrix->dimension = dimension;

    const int size = dimension * dimension;
    matrix->data = calloc(size, sizeof(double));
}

void matrix_fill_random(SquareMatrix input)
{
    for (int i = 0; i < input.dimension; ++i)
    {
        for (int j = 0; j < input.dimension; ++j)
        {
            input.data[i * input.dimension + j] = ((double)rand()) / RAND_MAX;
        }
    }
}

SquareMatrix matrix_multiply(SquareMatrix A, SquareMatrix B)
{
    const int dimension = A.dimension;
    assert(dimension == B.dimension && "B must have same dimension as A");

    SquareMatrix C;
    matrix_initialize(&C, dimension);

#pragma omp parallel for
    for (int i = 0; i < dimension; ++i)
    {
        for (int j = 0; j < dimension; ++j)
        {
            double acc = 0;
            for (int k = 0; k < dimension; ++k)
            {
                acc += A.data[i * dimension + k] * B.data[k * dimension + j];
            }
            C.data[i * dimension + j] = acc;
        }
    }

    return C;
}

void matrix_print(const SquareMatrix A)
{
    for (int i = 0; i < A.dimension; ++i)
    {
        for (int j = 0; j < A.dimension; ++j)
        {
            printf("%f ", A.data[i * A.dimension + j]);
        }
        puts("");
    }
}

void matrix_free(SquareMatrix *A)
{
    A->dimension = 0;
    free(A->data);
    A->data = NULL;
}

int main()
{
    const int dimension = 512;

    SquareMatrix A, B;
    matrix_initialize(&A, dimension);
    matrix_initialize(&B, dimension);

    srand(17);
    matrix_fill_random(A);
    matrix_fill_random(B);

    const double start = omp_get_wtime();
    SquareMatrix result = matrix_multiply(A, B);
    const double end = omp_get_wtime();
    const double duration = end - start;
    printf("Multiplication took %f seconds\n\n", duration);

    // printf("A:\n");
    // matrix_print(A);
    // printf("\nB:\n");
    // matrix_print(B);
    // printf("\nC = A * B:\n");
    // matrix_print(result);
    assert(result.dimension == A.dimension);

    puts("");

    matrix_free(&A);
    matrix_free(&B);
    matrix_free(&result);

    return 0;
}

File: openmp/matmul.f90
module matrix
   implicit none

   private

   public t_matrix, initialize_matrix, multiply_matrices, free_matrix

   type :: t_matrix
      integer :: size
      double precision, dimension(:, :), allocatable :: data
   end type

contains

   function initialize_matrix(size)
      integer, intent(in) :: size
      type(t_matrix) :: initialize_matrix
      type(t_matrix) :: mat

      mat%size = size
      allocate(mat%data(size, size))

      initialize_matrix = mat
   end function

   function multiply_matrices(a, b)
      type(t_matrix), intent(in) :: a, b
      type(t_matrix) :: multiply_matrices
      double precision, dimension(:, :), allocatable :: c
      integer :: n, i, j, k
      double precision :: acc

      n = a%size
      allocate(c(n, n))

      !$omp parallel private(acc) shared(c)
      !$omp do
      do i = 1, n
         do j = 1, n
            acc = 0
            do k = 1, n
               acc = acc + a%data(i, k) * b%data(k, j)
            end do
            c(i, j) = acc
         end do
      end do
      !$omp end parallel

      multiply_matrices%size = n
      multiply_matrices%data = c
   end function multiply_matrices

   subroutine free_matrix(a)
      type(t_matrix), intent(inout) :: a

      deallocate(a%data)
      a%size = 0
   end subroutine free_matrix

end module matrix


program matmul
   use matrix
   use omp_lib
   implicit none

   integer :: size = 512
   type(t_matrix) :: a, b, c
   double precision :: start_time, end_time, duration

   print *, "Initializing matrices with random data..."

   a = initialize_matrix(size)
   b = initialize_matrix(size)

   print *, "Multiplying matrices..."

   start_time = omp_get_wtime()
   c = multiply_matrices(a, b)
   end_time = omp_get_wtime()
   duration = end_time - start_time
   print '("Multiplication took ", F0.4, " seconds")', duration

   call free_matrix(a)
   call free_matrix(b)
   call free_matrix(c)

end program matmul
